tests/test_cli_compile.py

```diff
@@ -39,6 +41,55 @@
 )
 
 
+@pytest.fixture(scope="session")
+def installed_pip_version():
+    return get_pip_version_for_python_executable(sys.executable)
+
+
+@pytest.fixture(scope="session")
+def pip_produces_absolute_paths(installed_pip_version):
+    # in pip v24.3, new normalization will occur because `comes_from` started
+    # to be normalized to abspaths
+    return installed_pip_version >= Version("24.3")
+
+
+@dataclasses.dataclass
+class TestFilesCollection:
+    """
+    A small data-builder for setting up files in a tmp dir.
+
+    Contains a name for use as the ID in parametrized tests and contents.
+    'contents' maps from subpaths in the tmp dir to file content or callables
+    which produce file content given the tmp dir.
+    """
+
+    # the name for the collection of files
+    name: str
+    # static or computed contents
+    contents: dict[str, str | typing.Callable[[pathlib.Path], str]]
+
+    def __str__(self) -> str:
+        return self.name
+
+    def populate(self, tmp_path: pathlib.Path) -> None:
+        """Populate the tmp dir with file contents."""
+        for path_str, content in self.contents.items():
+            path = tmp_path / path_str
+            path.parent.mkdir(exist_ok=True, parents=True)
+            if isinstance(content, str):
+                path.write_text(content)
+            else:
+                path.write_text(content(tmp_path))
+
+    def get_path_to(self, filename: str) -> str:
+        """Given a filename, find the (first) path to that filename in the contents."""
+        return next(
+            k
+            for k in self.contents.keys()
+            if (k == filename) or k.endswith(f"/{filename}")
```

Comment:
`.keys()` is redundant in `dict`s. Also, it's not immediately obvious what the semantics of `k` is. I'll try to guess, though:
```suggestion
            stub_file_path
            for stub_file_path in self.contents
            if (stub_file_path == filename) or stub_file_path.endswith(f"/{filename}")
```

Reply:
Yeah, I think you came up with a good name for it. :+1: